### ISSUE TITLE
Raise warning about saving nn.Module in save_hyperparameters earlier

### ISSUE THREAD

**Issue by awaelchli** [Lightning-AI/lightning#15131]
## Proposed refactor

Raise warning about saving nn.Module in save_hyperparameters earlier

### Motivation

We have a warning when a user tries to "save" a nn.Module as hyperparameter here:

https://github.com/Lightning-AI/lightning/blob/8e83bfafcd92793910a76bbc41c58a91393da6fd/src/pytorch_lightning/utilities/parsing.py#L264-L267

This is to avoid redundantly big checkpoints and undesired behavior. The warning however is emitted after the deepcopy, and this can create memory issues or hangs and the warning will never be shown. 

### Pitch

Move the warning up so we can emit it earlier. 

### Additional context

Reported by user via social media.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

**Comment by cristi-zz** [comment]
Hi! Thank you for creating the issue out of a tweet! 

Here is a collab showing my issue. Interesting, sending a small nn.Module to the constructor will not reproduce the hanging part of the issue. Also, in collab, the "saving nn.Module" warning is shown. 

Unfortunately the machine where I work is not available right now, to check if this was shown. 100% possible to missed it!

https://colab.research.google.com/drive/19pbWe2LduqLRiGG64iVGpe2zp4Vitzuw?usp=sharing

Hope it helps!

**Comment by awaelchli** [comment]
@cristi-zz Thanks a lot. I confirm this is reproducible, and the warning is shown as expected. The program seems to hang when TensorBoard attempts to dump the hparams into a file here: https://github.com/Lightning-AI/lightning/blob/5dd7b6a4787819f12758a7d68d6c0cfdbed405c2/src/pytorch_lightning/core/saving.py#L405

I suppose yaml is not efficient in serializing the "big" resnet module. I'm sure it would eventually end or error out, but it will take a while. Not sure what we can do here except continue to show the warning. Perhaps a timeout?

**Comment by cristi-zz** [comment]
Thanks for looking into it! I doubt much can be done, except maybe a stronger lettered warning? I believe there are legitimate use cases when one would like to save a nn.Module. None cross my mind right now.

**Comment by awaelchli** [comment]
Yes agree. A simple use case could be a loss function, implemented as an nn.Module with no parameters.
